### PR TITLE
refactor(cli): cli deduplication in core

### PR DESCRIPTION
**Pull request by brandonspark**
## What:
This PR moves the CLI deduplication (based on `cli_unique_key`) to the Core engine.

## Why:
This makes the core output more in line with the cli output, and prevents duplication of the logic across pysemgrep and osemgrep.

## How:
The cli unique key is based on the rule metadata, so I had to move that to the pattern matches.

I also deleted the deduplication logic from the `pysemgrep` side.

## Test plan:
Tests still pass.

Note that I updated a bunch of snapshots. This is because of [this line](https://github.com/semgrep/semgrep/blob/39f95450a7d4d70e54c9edbd109bed8210a36889/cli/src/semgrep/core_runner.py#L771) in the pysemgrep core runner, which reorders the fields of the rule JSON before it enters the core engine.

When the core engine didn't used to deal with metadata, this means that `pysemgrep` would insert the metadata, in a non-sorted manner. Now, though, the core engine reports the sorted metadata, which is correct, as that's the metadata it receives. This changes the order in the snapshot data, though, so we need to change it to reflect the data the core engine is returning.

### CODE REVIEW

**Comment by brandonspark** [comment]
Test failure looks unrelated (I recompiled `pro` with latest `semgrep` and got the same Pro results)

**Comment by aryx** [comment]
Can we merge this?
Did you rebase?

**Comment by brandonspark** [comment]
@aryx this PR is finally done!

File: cli/tests/e2e-pro/snapshots/test_ci/test_dryrun/None/results.txt [hunk 1]
```diff
@@ -337,9 +337,9 @@ Would have sent findings and ignores blob: {
                 "semgrep.dev": {
                     "rule": {
                         "rule_id": "abcd",
-                        "version_id": "version1",
+                        "shortlink": "https://sg.run/abcd",
```
Comment (aryx) [inline cli/tests/e2e-pro/snapshots/test_ci/test_dryrun/None/results.txt:340]: not sure why we have those snapshot change which seems just to reorder fields.

**Review by aryx — APPROVED** [review]
Fantastic.

File: src/engine/Eval_generic.ml [hunk 2]
```diff
@@ -520,12 +520,13 @@ let bindings_to_env (config : Rule_options.t) ~file bindings =
     bindings
     |> List_.map_filter (fun (mvar, mval) ->
            let try_bind_to_exp e =
+             UCommon.(pr2 (spf "try bind %s" ([%show: AST_generic.expr] e)));
```
Comment (aryx) [inline src/engine/Eval_generic.ml:523]: we should not use pr2 ... we should use Logs.debug here.

File: src/engine/Eval_generic.ml [hunk 3]
```diff
@@ -602,6 +603,7 @@ let eval_opt env e =
       None
 
 let eval_bool env e =
+  UCommon.(pr2 (spf "code %s" ([%show: AST_generic.expr] e)));
```
Comment (aryx) [inline src/engine/Eval_generic.ml:606]: Logs.debug

File: src/reporting/Core_json_output.ml [hunk 4]
```diff
@@ -24,6 +24,17 @@ open Pattern_match
 module OutJ = Semgrep_output_v1_j
 module OutUtils = Semgrep_output_utils
 
+(*****************************************************************************)
+(* Types *)
+(*****************************************************************************)
+
+type rule_scan_source =
+  | Unchanged
```
Comment (aryx) [inline src/reporting/Core_json_output.ml:32]: Is this thing new? I don't see any similar code in pysemgrep.

Comment (brandonspark) [reply src/reporting/Core_json_output.ml:32]: it's part of the `RuleScanSource` enum here: https://github.com/semgrep/semgrep/blob/dc1477ed1e0b3bb63bad5257974f935c8989689b/cli/src/semgrep/constants.py#L55

File: src/engine/Eval_generic.ml (hunk 2 above)
Comment (brandonspark) [reply src/engine/Eval_generic.ml:523]: oops, these were accidentally included (I was debugging). Did you mean to merge with them in?

**Comment by aryx** [comment]
Please remove the UCommon.pr2 and actually add a rule for it in semgrep.jsonnet

**Pull request merged**